alf/environments/suite_babyai.py

```diff
@@ -0,0 +1,219 @@
+        'the',
+        'a',
+    ]
+
+    def __init__(self,
+                 env,
+                 max_instruction_length=32,
+                 one_token_per_step=False):
+        """
+        Args:
+            gym_env (gym.Env): An instance of OpenAI gym environment.
+            max_instruction_length (int): the maximum number of words of an instruction.
+            one_token_per_step (bool): If False, the whole instruction (word ID array)
+                is given in one step in the observation. If True, the word IDs are
+                given in the observation sequentially. Each step only one world ID
+                is given. Zeros are given for every steps after all the world IDs
+                are given.
+        """
+        super().__init__(env)
+
+        self._max_instruction_length = max_instruction_length
+        self._one_token_per_step = one_token_per_step
+
+        # the extra 1 is for padding
+        vocab_size = len(self.VOCAB) + 1
+
+        obs_space = {
+            'image':
+                env.observation_space['image'],
+            'direction':
+                gym.spaces.Discrete(4),
+            'mission':
+                gym.spaces.MultiDiscrete([vocab_size] * max_instruction_length)
+        }
+        if one_token_per_step:
+            obs_space['mission'] = gym.spaces.Discrete(vocab_size)
+
+        self.observation_space = gym.spaces.Dict(obs_space)
+
+        self._vocab = {'': 0}
+        for i, w in enumerate(self.VOCAB):
+            self._vocab[w] = i + 1
+        self._last_mission = ''
+        self._tokens = []
+        self._word_pattern = re.compile("([a-z]+)")
+
+    def _tokenize(self, instruction):
+        """Convert instruction string to a numpy array."""
+        tokens = self._word_pattern.findall(instruction.lower())
+        instr = np.array([self._vocab.get(token, 0) for token in tokens])
+        if np.amin(instr) == 0:
+            for token in tokens:
+                if token not in self._vocab:
+                    raise ValueError(
+                        "The instruction '%s' contains word "
+                        " out of vocabulary: %s" % (instruction, token))
+        if len(instr) < self._max_instruction_length:
+            instr = np.concatenate([
+                instr,
+                np.zeros([self._max_instruction_length - len(instr)],
+                         dtype=np.int64)
+            ])
+        elif len(instr) > self._max_instruction_length:
+            raise ValueError("The instruction is too long: %d" % len(instr))
+        return instr
+
+    def step(self, action):
+        obs, reward, done, info = self.env.step(action)
+        obs = self._transform_observation(obs)
+        return obs, reward, done, info
+
+    def reset(self, **kwargs):
+        obs = self.env.reset(**kwargs)
+        self._last_mission = ''
+        self._tokens = []
+        return self._transform_observation(obs)
+
+    def _transform_observation(self, observation):
```

Comment:
Should clarify that the original env gives the whole instruction at step=0 and gives '' for the remaining steps.